packages/relay/tests/lib/poller.spec.ts

```diff
@@ -0,0 +1,141 @@
+    ]);
+    const SINGLE_LINE = 'Called notifySubscriber with single line of log data!';
+    const tag = '{"event":"logs","filters":{"address":"0x23f5e49569A835d7bf9AefD30e4f60CdD570f225","topics":["0xc8b501cbd8e69c98c535894661d25839eb035b096adfde2bba416f04cc7ce987"]}}';
+
+    let ethImplStub: EthImpl;
+    let poller: Poller; 
+    let sandbox;
+
+    this.beforeEach(() => {
+ 
+        ethImplStub = sinon.createStubInstance(EthImpl);
+        ethImplStub.blockNumber.returns('0x1b177b');
+        ethImplStub.getLogs.returns(logs);
+
+        poller = new Poller(ethImplStub, logger);
+        
+        sandbox = sinon.createSandbox();
+
+    });
+
+    this.afterEach(() => {
+        sandbox.restore();
+    });
+
+
+    describe('Poller', () => {
+
+        it('should start polling', (done) => {
+            ethImplStub.blockNumber.returns('0x1b177b');
+            // eslint-disable-next-line @typescript-eslint/no-empty-function
+            const notifySubscriber = (tag, logs) =>{};
+            ethImplStub.getLogs.returns(logs);
+            const loggerSpy = sandbox.spy(logger, 'info');
+
+            poller.add(tag, notifySubscriber);
+            done();
+            
+            expect(poller.isPolling()).to.be.true;
+            expect(poller.hasPoll(tag)).to.be.true;
+            expect(loggerSpy.calledTwice).to.be.true;
+            expect(loggerSpy.getCall(0).args[0]).to.equal('Poller: Polling for {"event":"logs","filters":{"address":"0x23f5e49569A835d7bf9AefD30e4f60CdD570f225","topics":["0xc8b501cbd8e69c98c535894661d25839eb035b096adfde2bba416f04cc7ce987"]}}');
+            expect(loggerSpy.getCall(1).args[0]).to.equal('Poller: Starting polling');
+           
+        });
+
+        it('should stop polling', (done) => {
+            const loggerSpy = sandbox.spy(logger, 'info');            
+            poller.remove(tag);
+            done();
+
+            expect(poller.isPolling()).to.be.false;
+            expect(loggerSpy.callCount).to.equal(5);
```

Comment:
It's not clear from the logic why the tests should expect 5 calls here and 2 above.
I'm guessing you're using the logs as a proxy to measure progression through the stack.

The number here isn't as useful as the actual log check.
If removed a log and added it back it would make no different to this test.

I'd say remove the count check and verify the actual logs expected as you are below.
In which case you're missing 4 and 5

Reply:
This is a good catch.  That line of code should have failed, but the evaluations were not run because the done() was called before them.  I originally added the done() to finish the asynchronous function call.  I refactored both tests using done() to use the more robust async/await.  This exposed what I think is a bug.  @Ivo-Yankov please review.  This is what I was trying to explain when you showed us how to run the websockets server.  

In the poller.ts the hasPoll function should return true when it finds a matched tag in the Poll array.  I updated line 122 from:
```return !!this.polls.filter(p => p.tag !== tag).length;```
 that returns false when the tags match to:
```return !!this.polls.filter(p => p.tag === tag).length;```

The first test in the poller.spec.ts, lines 71 to 73 cover this.

Reply:
@ebadiere

Yes this is indeed a bug, I found it during acceptance test for subscription, when doing multiple subscriptions, and today I found the issue and then thought that this might be present on your Unit Tests, and then realized you found it and fix it there too... 
https://github.com/hashgraph/hedera-json-rpc-relay/issues/988